goldmane/proto/api.proto

```diff
@@ -47,15 +40,142 @@ message FlowRequest {
+
+  // AggregationInterval is the width of the time window in seconds across which to aggregate when generating
+  // Flows to return. This must be a multiple of 15.
+  int64 aggregation_interval = 7;
+}
+
+// StreamRequest defines a message to request a stream of aggregated Flows.
+message StreamRequest {
+  // StartTimeGt specifies the beginning of a time window from which to stream Flows. Flows
+  // will be streamed only if their start time occurs after the requested time.
+  int64 start_time_gt = 1;
+
+  // Filter allows specification of one or more criteria on which to filter the returned Flows.
+  Filter filter = 2;
+
+  // AggregationInterval defines both the frequency of streamed updates for each Flow, and the amount of time that FlowResult covers.
+  // It must always be 15s.
+  //
+  // Every AggregationInterval the server must send a FlowResult containing the aggregated data for that Flow from a
+  // time interval of width AggregationInterval.
+  //
+  // For a Flow that has continuous traffic, the server should send updates covering the range
+  // [now-2*AggregationInterval, now-AggregationInterval] so that the data is reasonably likely to be complete.
+  int64 aggregation_interval = 3;
+}
+
+message FilterHintsRequest {
+  // Type is type of Filter to query.
+  FilterType type = 1;
+
+  // Filter is a set of filter criteria used to narrow down returned results.
+  Filter filter = 2;
+
+  // StartTimeGt specifies the beginning of a time window with which to filter.
+  int64 start_time_gt = 3;
+
+  // StartTimeLt specifies the end of a time window with which to filter.
+  int64 start_time_lt = 4;
+}
+
+message FilterHint {
+  string value = 1;
+}
+
+// FilterType specifies which fields on the underlying Flow data to collect.
+enum FilterType {
+  FilterTypeDestName = 0;
+  FilterTypeSourceName = 1;
+  FilterTypeDestNamespace = 2;
+  FilterTypeSourceNamespace = 3;
+  FilterTypePolicyTier = 4;
+}
+
+// FlowResult wraps a Flow object with additional metadata.
+message FlowResult {
+  // ID is an opaque integer value ID that can be used to identify a Flow, and is 1:1 with the FlowKey.
+  // Note that this ID is not valid across server restarts. Its primary use-case is for correlating FlowResult
+  // updates from a Stream request.
+  int64 id = 1;
+
+  // The Flow object itself.
+  Flow flow = 2;
+}
+
+// Filter defines criteria for selecting a set of Flows based on their parameters.
+message Filter {
+  string source_name = 1;
+  string source_namespace = 2;
+  string dest_name = 3;
+  string dest_namespace = 4;
+  string protocol = 5;
+  int64 dest_port = 6;
+  PolicyMatch policy = 7;
+}
+
+message SortOption {
+  // SortBy delcares the field by which to sort.
+  SortBy sort_by = 1;
+}
+
+// PolicyMatch defines criteria for matching one or more policy rules within a Flow's
+// policy trace.
+message PolicyMatch {
+  PolicyKind Kind = 1;
+  string Tier = 2;
+  string Namespace = 3;
+  string Name = 4;
+  string Action = 5;
```

Comment:
Just to confirm, to see if a flow was allowed or denied you need to traverse the policies?

Does this always align with if the flow was actually denied or allowed, like with staged policy?

Reply:
Nope, there's a top-level Action field in the FlowKey as well that contains the final action and is what this matches against. 

I'm not actually sure what the top-level Action field will display for staged policy actions - I suspect it only shows the "enforced" action and for staged you do need to look at the pending_policies array.

Reply:
Ahh right, the key.